flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java

```diff
@@ -238,6 +248,12 @@ public void setup() throws IOException {
 		checkState(this.bufferPool == null, "Bug in input gate setup logic: Already registered buffer pool.");
 		setupChannels();
 
+		//TODO we can further judge the condition whether the type is file-based or mmap-based
+		if (consumedPartitionType.isBlocking()) {
```

Comment:
I would simply make this a simple lazy-created MemorySegement, without any logic trying to determine what the shuffle type is. If this is fetched only if needed (see using `Supplier<MemorySegment>` above) then that will be good enough.

I think it would also be fine if this was an unpooled segment, not obtained from the `memorySegmentProvider` but just created on first access. Would make things simpler, you don't have to worry about recycling, races between creation and early disposal, etc.

Reply:
Agree, it seems better to make it lazy-created and unpooled segment.

Reply:
I thought of another issue to use `Supplier<MemorySegment>` way, that means for every call of `LocalInputChannel#getNextBuffer` we need to allocate a new segment for it (no matter with pooled or unpooled segment). My previous way tried to allocate only one segment for always reuse in all the ``LocalInputChannel#getNextBuffer` calls.

But I am fine with the proposal of using unpooled segment here, because it indeed simplify some aspects as you mentioned above.